extensions/interactions/DragAndDropSort/directives/DragAndDropSortValidationService.js

```diff
@@ -0,0 +1,117 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Validator service for the drag and drop sorting interaction.
+ */
+
+oppia.factory('DragAndDropSortValidationService', [
+  'baseInteractionValidationService', 'WARNING_TYPES',
+  function(baseInteractionValidationService, WARNING_TYPES) {
+    return {
+      getCustomizationArgsWarnings: function(customizationArgs) {
+        var warningsList = [];
+
+        baseInteractionValidationService.requireCustomizationArguments(
+          customizationArgs, ['choices']);
+
+        var areAnyChoicesEmpty = false;
+        var areAnyChoicesDuplicated = false;
+        var seenChoices = [];
+        var numChoices = customizationArgs.choices.value.length;
+
+        for (var i = 0; i < numChoices; i++) {
+          var choice = customizationArgs.choices.value[i];
+          if (choice.trim().length === 0) {
+            areAnyChoicesEmpty = true;
+          }
+          if (seenChoices.indexOf(choice) !== -1) {
+            areAnyChoicesDuplicated = true;
+          }
+          seenChoices.push(choice);
+        }
+
+        if (areAnyChoicesEmpty) {
+          warningsList.push({
+            type: WARNING_TYPES.CRITICAL,
+            message: 'Please ensure the choices are nonempty.'
+          });
+        }
+
+        if (areAnyChoicesDuplicated) {
+          warningsList.push({
+            type: WARNING_TYPES.CRITICAL,
+            message: 'Please ensure the choices are unique.'
+          });
+        }
+
+        return warningsList;
+      },
+      getAllWarnings: function(
```

Comment:
Add validation checks for rules. For e.g. if I have the first rule "is equal to [[1], [2, 3]] order with one element in wrong position" and second rule "is equal to [[1], [2], [3]] order" then second rule can never be reached.

Reply:
I agree. This is a very specific case. I think "is equal to" rule can be followed by "is equal to with one item at incorrect position" and vice-versa except for this particular case.

Reply:
Yep, its a special case. You should also add a validation check for "is equal to this order with one item in wrong place" when there are no multiple items occupying same rank, i.e. each rank has exactly one item. In that case, we simply can't have an input configuration where only one item is in wrong place and rule would never ever fire

Reply:
Yes, but how can we be so sure that the learner will not do so? Probably a question may have an answer with no items at same position. But the creator may want to give some feedback with the learner puts one item at same position.

Reply:
Ah, yes. I got your point. I didn't consider that possibility. Ignore my previous comment, then.

Reply:
Done